shared/src/metabase/shared/util/internal/time.cljs

```diff
@@ -0,0 +1,150 @@
+(ns metabase.shared.util.internal.time
+  "CLJS implementation of the time utilities on top of Moment.js.
```

Comment:
The idea was to move the shared logic to Clojure from JS. Having the frontend-only logic in Clojure will make FE engineers' lives slightly harder because we will have to deal with this code while not being very productive with Clojure. Is this a necessity?

Reply:
I thought we will inject a platform-specific date formatting function into the shared code and all the code which deals with Moment will remain in the JS world 🤔

Reply:
> ... inject a platform-specific date formatting function into the shared code and all the code which deals with Moment will remain in the JS world 🤔

I think that would make it difficult to test that the code produces the same results on both platforms.

Reply:
I draw a distinction between FE-only logic and using each platform's date/time library.

The properly FE-only logic (`jsx` output, a couple of `column`-related things) still lives in `frontend/src/metabase/lib/formatting/date.tsx`.

This Clojure(Script) code produces identically formatted dates on both platforms. Most of the code is shared, but there's some guts that lean on the respective platforms' date/time libraries to do the basic parsing, formatting and date arithmetic. We can't share that without writing it all, and that's a huge project.

Moving any code from JS to Clojure is going to make it an uphill change for FE folks. The hope with this metabase-lib project is that the sharing will lead to more closely-matching behavior on both sides and less total maintenance.

Are there specific bits in the Clojure that feel like purely FE concerns?

Reply:
I would consider any code that imports moment.js to be the FE code because it cannot be used on the backend and exists only for the FE. I see that there are not that many cljs like that so given the arguments you and Tamas provided let's give it a shot

Reply:
Theoretically, we could write a JS object full of the various date/time functions we need, and pass that to the shared code. That would keep the Moment dep and details contained to the FE code.

On the downside, it would reduce the sharing somewhat and make the flow of control harder to follow. It's also tricky to write a test suite that runs against both environments and so makes sure the output is identical. (Without duplicating that object of Moment functions or invoking GraalVM like static viz does.)

Given that the Moment-specific parts of this code in `internal/time.cljs` are pretty low level, they're unlikely to change much, if ever. (Even more so since Moment is end-of-life and only getting bug fixes.) The parts of this code that do change are things like your recent PR to add weekdays, which touches only the shared code.